src/asm/x86/transform/inverse.rs

```diff
@@ -80,80 +77,80 @@ pub trait InvTxfm2D: native::InvTxfm2D {
   {
     debug_assert!(bd == 8);
 
-    // TODO: Support more than Block_4x4 when SSSE3 and AVX2 match.
-    debug_assert!(Self::W == 4 && Self::H == 4);
-
-    let mut coeff16: AlignedArray<[i16; 4 * 4]> =
+    // 64x only uses 32 coeffs
+    let coeff_w = Self::W.min(32);
+    let coeff_h = Self::H.min(32);
+    let mut coeff16: AlignedArray<[i16; 32 * 32]> =
       AlignedArray::uninitialized();
 
     // Transpose the input.
-    assert!(input.len() >= 4 * 4);
-    for j in 0..4 {
-      for i in 0..4 {
-        coeff16.array[i * 4 + j] = input[j * 4 + i] as i16;
+    // TODO: should be possible to remove changing how coeffs are written
+    assert!(input.len() >= coeff_w * coeff_h);
+    for j in 0..coeff_h {
+      for i in 0..coeff_w {
+        coeff16.array[i * coeff_h + j] = input[j * coeff_w + i] as i16;
       }
     }
 
     let stride = output.plane_cfg.stride as isize;
 
     // perform the inverse transform
-    (match tx_type {
-      TxType::DCT_DCT => rav1e_inv_txfm_add_dct_dct_4x4_ssse3,
-      TxType::IDTX => rav1e_inv_txfm_add_identity_identity_4x4_ssse3,
-      TxType::DCT_ADST => rav1e_inv_txfm_add_adst_dct_4x4_ssse3,
-      TxType::ADST_DCT => rav1e_inv_txfm_add_dct_adst_4x4_ssse3,
-      TxType::DCT_FLIPADST => rav1e_inv_txfm_add_flipadst_dct_4x4_ssse3,
-      TxType::FLIPADST_DCT => rav1e_inv_txfm_add_dct_flipadst_4x4_ssse3,
-      TxType::V_DCT => rav1e_inv_txfm_add_identity_dct_4x4_ssse3,
-      TxType::H_DCT => rav1e_inv_txfm_add_dct_identity_4x4_ssse3,
-      TxType::ADST_ADST => rav1e_inv_txfm_add_adst_adst_4x4_ssse3,
-      TxType::ADST_FLIPADST => rav1e_inv_txfm_add_flipadst_adst_4x4_ssse3,
-      TxType::FLIPADST_ADST => rav1e_inv_txfm_add_adst_flipadst_4x4_ssse3,
-      TxType::FLIPADST_FLIPADST => {
-        rav1e_inv_txfm_add_flipadst_flipadst_4x4_ssse3
-      }
-      TxType::V_ADST => rav1e_inv_txfm_add_identity_adst_4x4_ssse3,
-      TxType::H_ADST => rav1e_inv_txfm_add_adst_identity_4x4_ssse3,
-      TxType::V_FLIPADST => rav1e_inv_txfm_add_identity_flipadst_4x4_ssse3,
-      TxType::H_FLIPADST => rav1e_inv_txfm_add_flipadst_identity_4x4_ssse3,
-    })(
+    Self::match_tx_type_ssse3(tx_type)(
       output.data_ptr_mut() as *mut _,
       stride,
       coeff16.array.as_ptr(),
-      (4 * 4) as i32,
+      (coeff_w * coeff_h) as i32,
     );
   }
 }
 
 macro_rules! impl_itx_fns {
   // Takes a 2d list of tx types for W and H
-  ([$([$(($ENUM:pat, $TYPE1:ident, $TYPE2:ident)),*]),*], $W:expr, $H:expr,
-   $OPT:ident) => {
+  ([$([$(($ENUM:pat, $TYPE1:ident, $TYPE2:ident)),*]),*], $W:literal, $H:literal) => {
     paste::item! {
       // For each tx type, declare an function for the current WxH
-      $(
+      extern {
         $(
-          extern {
+          $(
             // Note: type1 and type2 are flipped
-            fn [<rav1e_inv_txfm_add_ $TYPE2 _$TYPE1 _$W x $H _$OPT>](
+            fn [<rav1e_inv_txfm_add_ $TYPE2 _$TYPE1 _$W x $H _avx2>](
```

Comment:
Wouldn't be enough to make OPT a `,-list` and expand it in the macro?

Reply:
I couldn't work out the right syntax for making the Cartesian product of parameter-list and opt-list. Any example of a macro like this would be helpful.